src/eventual/writer.rs

```diff
@@ -0,0 +1,56 @@
+use futures::{channel::oneshot::Receiver, future::Shared};
+
+use super::*;
+use crate::error::Closed;
+use futures::FutureExt;
+use std::sync::{Arc, Weak};
+
+pub struct EventualWriter<T>
+where
+    T: Value,
+{
+    state: Weak<SharedState<T>>,
```

Comment:
I assume that if this is a `Weak` reference, something else owns the `SharedState<T>`? Or is `Weak` used so that if all subscribers/readers are dropped, they drop their `SharedState<T>` references and after this, the weak reference will contain `None`?

Reply:
It's `Weak` so that the `SharedState` drops when all the readers drop. Once the readers are dropped the writer can be notified. If the Writer held onto the SharedState strongly then it would never be notified. In this way the automatic cancellation is bi-directional. (The writer stops if there is no possibility that values will be read, and a reader stops when there is no possibility of further writes)